### (1) Gathering data

In [1]:
import pandas as pd
import numpy as np
import requests
import io
import json
import re
import os
import matplotlib.pyplot as plt
%matplotlib inline
from PIL import Image
from io import BytesIO

In [2]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name = 'image_predictions.tsv'
if not os.path.exists(file_name):
    urlData = requests.get(url).content
    with open(file_name, mode='wb') as file:
        file.write(urlData)
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

In [4]:
# read twitter_api.py file to understand the code
with open('twitter_api.py') as f:
    twitter_api = f.readlines()
for line in twitter_api:
    print(line)

import tweepy

from tweepy import OAuthHandler

import json

from timeit import default_timer as timer



# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file

# These are hidden to comply with Twitter's API terms and conditions

consumer_key = 'HIDDEN'

consumer_secret = 'HIDDEN'

access_token = 'HIDDEN'

access_secret = 'HIDDEN'



auth = OAuthHandler(consumer_key, consumer_secret)

auth.set_access_token(access_token, access_secret)



api = tweepy.API(auth, wait_on_rate_limit=True)



# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:

# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to

# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv

# NOTE TO REVIEWER: this student had mobile verification issues so the following

# Twitter API code was sent to this student from a Udacity instructor

# Tweet IDs for which to gather additional data via Twitter's API

tweet_ids = df_1.tweet_id.valu

In [5]:
with open('tweet.json', 'r') as json_file:
    file = '[' + json_file.read().replace('}\n{', '},\n{') + ']'
    tweet_json = json.loads(file)
tweet_info = pd.DataFrame(tweet_json)

### (2) Assess

In [6]:
tweet_info.head(2)

contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           39467   
1  {'media': [{'id': 892177413194625024, 'id_str'...           33819   

   favorited                                          full_text   geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None   
1      False  This is Tilly. She's just checking pup on you....  None   

                         ...                          \
0                        ...                           
1                        ...                           

   possibly_sensitive_appealable quoted_status quoted_status_id  \
0                          False           NaN              NaN   
1                          False           NaN              NaN   

   quoted_status_id_str retweet_count  retweeted retweeted_status  \
0                   NaN          8853      False              NaN   
1                   NaN          6514      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[2 rows x 31 columns]

In [7]:
tweet_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
id_str                           2354 non-null object
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obj

In [8]:
tweet_info.describe()

favorite_count            id  in_reply_to_status_id  \
count     2354.000000  2.354000e+03           7.800000e+01   
mean      8080.968564  7.426978e+17           7.455079e+17   
std      11814.771334  6.852812e+16           7.582492e+16   
min          0.000000  6.660209e+17           6.658147e+17   
25%       1415.000000  6.783975e+17           6.757419e+17   
50%       3603.500000  7.194596e+17           7.038708e+17   
75%      10122.250000  7.993058e+17           8.257804e+17   
max     132810.000000  8.924206e+17           8.862664e+17   

       in_reply_to_user_id  quoted_status_id  retweet_count  
count         7.800000e+01      2.900000e+01    2354.000000  
mean          2.014171e+16      8.162686e+17    3164.797366  
std           1.252797e+17      6.164161e+16    5284.770364  
min           1.185634e+07      6.721083e+17       0.000000  
25%           3.086374e+08      7.888183e+17     624.500000  
50%           4.196984e+09      8.340867e+17    1473.500000  
75%           4.196984e+09      8.664587e+17    3652.000000  
max           8.405479e+17      8.860534e+17   79515.000000

In [9]:
twitter_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [ ]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [ ]:
twitter_archive.text[200]

In [ ]:
twitter_archive.source.value_counts()

In [ ]:
twitter_archive.expanded_urls.sample(5)

In [ ]:
twitter_archive.rating_numerator.value_counts()

In [ ]:
twitter_archive.rating_denominator.value_counts()

In [ ]:
twitter_archive[twitter_archive['rating_numerator'] > 20][['rating_numerator','rating_denominator']]

In [ ]:
twitter_archive[(twitter_archive['doggo']=='None')&(twitter_archive['floofer']=='None')&(twitter_archive['pupper']=='None')&(twitter_archive['puppo']=='None')].shape[0]

In [ ]:
image_predictions.head(3)

In [ ]:
image_predictions.info()

In [ ]:
image_predictions.img_num.value_counts()

In [ ]:
twitter_archive[twitter_archive['name'].str.len() < 3]['name'].value_counts()

####  quality issues
- in `twitter_archive` dataset, `tweet_id` data type is int instead of str.
- there are many missing values in `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, and `retweeted_status_timestamp`columns in `twitter_archive` dataset, we can remove these columns.
- `name` column in `twitter_archive` has strange names like(a, an, the, o)
- `timestamp` column at `twitter_archive` dataset data type is object instead of datetime.
- `text` column at `twitter_archive` dataset has a link to the tweet.
- `source` column in `twitter_archive` dataset has html tag, we can extract the link from it.
- `expanded_urls` column in `twitter_archive` has null values.
- in `image_predictions` dataset, `tweet_id` data type is int instead of str.
- `rating_numerator` column has outliers, and it's better to be float.
- `rating_denominator` column has values not 10, I want all values equal 10 to drop this column.
- remove the retweeted twitter

#### tidiness issues
- `doggo`,`floofer`,`pupper`,`puppo` columns should be one column `dog_kind`.
- the three datasets should merged to be only one.
- just 3 columns needed in `tweet_info` dataset (id_str,favorite_count,retweet_count)
- create new colmun (`dog_species`) to store dog species that is the highest probability to be correct.

#### make a copy from the three dataset to clean.

In [ ]:
twitter_archive_clean = twitter_archive.copy()
tweet_info_clean = tweet_info.copy()
image_predictions_clean = image_predictions.copy()

### (3) clean

####  quality


#### define
convert the data type of `tweet_id` in `twitter_archive` dataset to str.

#### code

In [ ]:
twitter_archive_clean['tweet_id'] = twitter_archive_clean['tweet_id'].astype(str)

#### test

In [ ]:
twitter_archive_clean.info()

#### define
- drop `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_user_id`, and `retweeted_status_timestamp`columns in `twitter_archive` dataset.

#### code

In [ ]:
twitter_archive_clean.drop(['in_reply_to_status_id',
                            'in_reply_to_user_id',
                            'retweeted_status_user_id',
                            'retweeted_status_timestamp'], axis=1, inplace=True)

#### test

In [ ]:
twitter_archive_clean.info()

#### define 
convert the data type of `timestamp` column at `twitter_archive` dataset to datetime.

#### code

In [ ]:
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean['timestamp'])

#### test

In [ ]:
twitter_archive_clean.info()

#### define 
extract the link from `text` column at `twitter_archive` dataset and then remove this column and put the link in a new column called `tweet_url`.

#### code

In [ ]:
twitter_archive_clean['tweet_url'] = twitter_archive_clean['text'].str.extract('(https://t.co/[A-z0-9_]+)')


#### test

In [ ]:
twitter_archive_clean.info()

In [ ]:
twitter_archive_clean['tweet_url'].head()

#### define 
extract the link from `source` column in `twitter_archive` dataset

#### code

In [ ]:
twitter_archive_clean['source'] = twitter_archive_clean['source'].str.extract('(http(s)?://[a-z]+\.[a-z]+\.?/?([a-z]+)?/?([a-z]+)?/?([a-z]+)?)')[0]

#### test

In [ ]:
twitter_archive_clean.source.unique()

#### define 
drop null values in expanded_urls.

#### code

In [ ]:
twitter_archive_clean = twitter_archive_clean.dropna(subset=['expanded_urls'])

#### test

In [ ]:
twitter_archive_clean['expanded_urls'].isna().sum()

In [ ]:
twitter_archive_clean.info()

#### define
convert the data type of `tweet_id` column in `image_predictions` dataset to str

#### code

In [ ]:
image_predictions_clean['tweet_id'] = image_predictions_clean['tweet_id'].astype(str)

#### test

In [ ]:
image_predictions_clean.info()

#### define
- make all values in `rating_denominator` column equal 10 with respecting the corresponding `rating_numerator`
- there is a correcting rating in `text` column, we can extract it using regular expression.

#### code

In [ ]:
twitter_archive_clean[twitter_archive_clean['rating_denominator'] != 10][['text','rating_numerator','rating_denominator']]

In [ ]:
pattern = "\s*(\d+([.]\d+)?([/]\d+))"
def ratings(x):
    return [m.group() for m in re.finditer(pattern, x)]

def clean(rating_denominator, rating_numerator, text):
    if rating_denominator > 10 and rating_denominator%10 == 0:
        rating_numerator = round(rating_numerator/rating_denominator*10.0, 1)
        rating_denominator = rating_denominator/rating_denominator*10
        return pd.Series([rating_denominator, rating_numerator])
    else:
        rating = ratings(text)
        if rating[-1].split('/')[1] == '10':
            rating_denominator = 10
            rating_numerator = round(float(rating[-1].split('/')[0]), 1)
            return pd.Series([rating_denominator, rating_numerator])
        else:
            rating_numerator = round(rating_numerator/rating_denominator*10.0, 1)
            rating_denominator = rating_denominator/rating_denominator*10
            return pd.Series([rating_denominator, rating_numerator])
    return pd.Series([rating_denominator, rating_numerator])

In [ ]:
twitter_archive_clean[['rating_denominator', 'rating_numerator']] = twitter_archive_clean.apply(lambda row: clean(row['rating_denominator'], row['rating_numerator'], row['text']), axis=1)

#### test

In [ ]:
twitter_archive_clean['rating_denominator'].unique()

#### Notice:
- now all denominators is 10, so `rating_denominator` column is useless, we can remove it.
- also `text` column is usless now, we can remove it.

In [ ]:
twitter_archive_clean.drop('rating_denominator', axis=1, inplace=True)
twitter_archive_clean.drop('text', axis=1, inplace=True)

#### define
- `rating_numerator` column has outliers.
- convert `rating_numerator` data type to float.
- let's make any value exceeds 15 to be 15, and any value less than 3 to be 3.
- because we removed `rating_denominator` column, I will change `rating_numerator` column name to be `rating` only 

In [ ]:
twitter_archive_clean['rating_numerator'] = twitter_archive_clean['rating_numerator'].astype(float)

#### code

In [ ]:
twitter_archive_clean['rating_numerator'].value_counts()

In [ ]:
twitter_archive_clean.loc[twitter_archive_clean['rating_numerator'] > 15,'rating_numerator'] = 15
twitter_archive_clean.loc[twitter_archive_clean['rating_numerator'] < 3,'rating_numerator'] = 3

In [ ]:
twitter_archive_clean.rename(columns={'rating_numerator':'rating'}, inplace=True)

#### test

In [ ]:
twitter_archive_clean['rating'].value_counts()

#### define
replace invalid names with None

#### code

In [ ]:
# notice: correct names start with capital characters,
#so we will replace all names that don't start with capital character
twitter_archive_clean['name'][twitter_archive_clean['name'].str.match('[a-z]+')] = 'None'

#### test

In [ ]:
twitter_archive_clean['name'][twitter_archive_clean['name'].str.match('[a-z]+')]

#### define
remove the retweeted twitter

#### code

In [ ]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean['retweeted_status_id'].isna()]

#### test

In [ ]:
twitter_archive_clean['retweeted_status_id'].unique()

#### now we can drop `retweeted_status_id` column

In [ ]:
twitter_archive_clean.drop('retweeted_status_id', axis=1, inplace=True)

In [ ]:
twitter_archive_clean.info()

In [ ]:
(twitter_archive_clean[twitter_archive_clean.expanded_urls.isna()].index == twitter_archive_clean[twitter_archive_clean.tweet_url.isna()].index ).all()

In [ ]:
twitter_archive_clean.expanded_urls.head()

#### tidiness
- `doggo`,`floofer`,`pupper`,`puppo` columns should be one column `dog_kind`.
- the three datasets should merged to be only one.
- just 3 columns needed in `tweet_info` dataset (id_str,favorite_count,retweet_count)

#### define
 make the 4 stages of dog in one column instead of 4.

In [ ]:
twitter_archive_clean['stage'] = twitter_archive_clean.apply(lambda row: ','.join([row['doggo'], row['floofer'], row['pupper'], row['puppo']]), axis=1).str.replace(',None','')
twitter_archive_clean['stage'] = twitter_archive_clean['stage'].str.replace('None,','')

In [ ]:
twitter_archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1, inplace=True)

#### test

In [ ]:
twitter_archive_clean.info()

In [ ]:
twitter_archive_clean[twitter_archive_clean.stage.str.len()>7].stage

#### define
just 3 columns needed in `tweet_info` dataset (id_str,favorite_count,retweet_count).

#### code

In [ ]:
tweet_info_clean = tweet_info_clean[['id_str', 'favorite_count', 'retweet_count']]
tweet_info_clean.rename(columns={'id_str':'tweet_id'}, inplace=True)

#### test

In [ ]:
tweet_info_clean.head()

#### define
concatenate the three datasets to be one.

#### code

In [ ]:
twitter_archive_clean = twitter_archive_clean.merge(tweet_info_clean,
                                                    left_on='tweet_id',
                                                    right_on='tweet_id')
twitter_archive_clean = twitter_archive_clean.merge(image_predictions_clean,
                                                    left_on='tweet_id',
                                                    right_on='tweet_id')

#### test

In [ ]:
twitter_archive_clean.head(3)

In [ ]:
twitter_archive_clean.shape

In [ ]:
# remove observations that aren't dogs
twitter_archive_clean = twitter_archive_clean[((twitter_archive_clean['p1_dog'] == True) |
                                               (twitter_archive_clean['p2_dog'] == True) |
                                               (twitter_archive_clean['p3_dog'] == True))]

In [ ]:
twitter_archive_clean.info()

In [ ]:
twitter_archive_clean.head()

In [ ]:
twitter_archive_clean.p1.nunique()

#### define
create new colmun (`dog_species`) to store dog species that is the highest probability to be correct.

#### code

In [ ]:
def dog_species(row):
    if row.p1_dog:
        return row.p1
    elif row.p2_dog:
        return row.p2
    else:
        return row.p3
twitter_archive_clean['dog_species'] = twitter_archive_clean.apply(lambda r:  dog_species(r), axis=1)

#### test

In [ ]:
twitter_archive_clean[['p1', 'p1_dog', 'p2', 'p2_dog', 'p3', 'p3_dog', 'dog_species']]

### Storing, Analyzing, and Visualizing Data

In [ ]:
twitter_archive_clean.to_csv('twitter_archive_master.csv', index=False)

In [ ]:
df = pd.read_csv('twitter_archive_master.csv')
df.head()

### what is the most 10 frequently dogs species?

In [ ]:
dog_species_df = df.groupby(['dog_species'], as_index=False)['tweet_id'].count().rename(columns={'tweet_id':'count'})
most_dog_species_df = dog_species_df.sort_values('count', ascending=False).head(10)
most_dog_species_df

In [ ]:
plt.bar(most_dog_species_df.dog_species, most_dog_species_df['count'])
plt.xticks(rotation=45, ha='right')
plt.xlabel('Dog species')
plt.ylabel('Number of tweets')
plt.title('Dog species vs Number of tweets')

### what is the least 10 frequently dogs species?

In [ ]:
least_dog_species_df = dog_species_df.sort_values('count', ascending=True).head(10)
least_dog_species_df

### most rating

In [ ]:
df.rating.value_counts()

In [ ]:
plt.hist(df.rating, bins=13)
plt.xlabel('Rating')
plt.ylabel('Count')
plt.title('Count of each Rating')

In [ ]:
df.head()

### which is the tweet that has the most favorite_count?

In [ ]:
max_favorite_count = df[df.favorite_count == df.favorite_count.max()]
max_favorite_count

### the image with the highest favorite_count

In [ ]:
r = requests.get(max_favorite_count.jpg_url.values[0])
i = Image.open(BytesIO(r.content))
i

### which is the tweet that has the most retweet_count?

In [ ]:
max_retweet_count = df[df.retweet_count == df.retweet_count.max()]
max_retweet_count

### the image with the highest retweet_count

In [ ]:
r = requests.get(max_retweet_count.jpg_url.values[0])
i = Image.open(BytesIO(r.content))
i

In [ ]:
most_dog_species_df.dog_species.values

### insights

- the most 10 frequently dog species is ('golden_retriever', 'Labrador_retriever', 'Pembroke', 'Chihuahua', 'pug', 'toy_poodle', 'chow', 'Pomeranian', 'Samoyed', 'malamute').


- the least 10 frequently dog species is ('Scotch_terrier', 'standard_schnauzer', 'silky_terrier', 'clumber', 'Japanese_spaniel', 'Irish_wolfhound', 'EntleBucher', 'Bouvier_des_Flandres', 'groenendael', 'black-and-tan_coonhound').


- the tweet that have the tweet_id = '822872901745569793', has the most favorite_count that equal 132810 	


- the tweet that have the tweet_id = '744234799360020481', has the most retweet_count that equal 131075